[MDS-5467] - Download Document Versions

matbusby-fw opened the pull request:
[MDS-5467](https://bcmines.atlassian.net/browse/MDS-5467)
[MDS-5346](https://bcmines.atlassian.net/browse/MDS-5346)
[MDS-5471](https://bcmines.atlassian.net/browse/MDS-5471)

Included the fixes for 5346 and 5471 as they touched on the same elements modified in this PR

## Objective 

- Reworked the saving of versions (it looked like it was working with my previous changes, but wasn't actually saving the versionId)
- updated the actionless network call to download a file, to optionally include a versionId
- updated the documant manager api to accept the versionId on a download and utilize it if it exists.
- Added a MineDocumentVersion class to `document.ts`. Previously versions were using the same model as documents, but they do not adhere to the same structure.
- Made spatial files replacable in the `ReplaceDocumentModal`
- Removed actions from documents that do not yet have a mine_document_guid

matbusby committed:
Commit message:
Add version handling for document download and refactor models

This commit adds handling for document versions while downloading a document, and enhances it to download a specific version if provided. The 'document_manager_version_guid' field was introduced to MineDocument model for maintaining version ids. The downloadFileFromDocumentManager utility has also been modified for conditional usage of this new field. Moreover, the MineDocument model has been refactored to have more clarity, and a new model 'MineDocumentVersion' has been created to handle different versions of the document. Finally, updates were made for integration with tusd_hooks.py and other related services to facilitate the version download. These changes were necessary to support different versions of a document in our application.

taraepp commented on services/core-web/common/models/documents/document.ts:
```diff
@@ -150,25 +200,13 @@ export class MajorMineApplicationDocument extends MineDocument {
     );
   }
 
-  protected makeChild(params: any, constructorArgs: any) {
```
I think there will be some difficulty in making versions a different type from the latest version. The file permissions are based on the type, and it keeps the properties consistent for document tables. If it works, that'd be a nice bit of clean-up, but I would be surprised if the document tables still have the right column data and allowed actions based on user roles.

matbusby-fw replied:
The data in the columns is still displaying correctly for the versions actually 😄 

They have a handful of common fields between MineDocuments and MineDocumentVersions but passing the version info into a MineDocument was mostly just ending up with a bunch of undefined fields.  Since versions should always conform to the same structure, I think it makes sense to simplify them like this.

I did add the allowed actions for versions, since they should always just be download and or view.

taraepp replied:
Ahh, that's right, we took out the File Location column that displays the major_mine_application_type_code, so it wouldn't produce an error with the columns we're using. If we're not planning on displaying versions with columns that are specific to the type then this shouldn't be a problem.

Might still suggest passing the view/download permission from the parent, so that we'd be able to restrict that based on permission (should be just as easy). I thought that the child data still wanted a unique key though- are you not getting a "Each record in table should have a unique key prop" message? No issues sorting? <-- that's when I ran into weird bugs

matbusby-fw replied:
Hmmm, let me check the key issue.

Sure, we can pass the view/download permission from the parent.  Do we ever restrict this though?  I thought we only restricted the other functions.

matbusby committed:
Update key references and add allowed actions for documents

The `key` field in the `document.spec.ts` test suite was replaced with `document_manager_version_guid` for `mineDocumentRecord` and its versions. This change ensures consistency in managing and referencing versions of document records. Also, a new property `allowed_actions` was incorporated into the document model, enabling each document to carry information about the permissible actions. A test suite relating to `MajorMineApplicationDocument` versions was also removed, its relevance was no longer present after the changes. A console.log statement was added to `DocumentTable.tsx` for debugging purposes.

matbusby committed:
update snap

matbusby committed:
match document.ts

matbusby committed:
match document.ts

matbusby committed:
match document.ts

matbusby committed:
match document.ts

matbusby committed:
match document.ts

matbusby committed:
Improve document handling and UI enhancement

Updated the list of accepted file types in document modals to include SPATIAL files. Allowing SPATIAL files ensures users have a more extensive range of data types to choose from when uploading or replacing files in the document modals.

Adjusted the document action dropdown in CoreTableCommonColumns.tsx. The dropdown now only appears when there are items in the list to prevent UI inconsistencies.

Refined the filtering of actions in DocumentTable.tsx for smoother performance, and removed a console.log used for debugging.

Ensured only documents with a valid mine_document_guid can have their actions modified in document.ts.

Lastly, adjusted the indentation in ReplaceDocumentModal.tsx for better readability and consistency throughout the codebase.

matbusby committed:
updated ReplaceDocumentModal.spec.tsx.snap

matbusby committed:
updated ReplaceDocumentModal.spec.tsx.snap
